projects/plugins/jetpack/modules/masterbar/admin-menu/class-base-admin-menu.php

```diff
@@ -161,12 +164,41 @@ public function update_submenus( $slug, $submenus_to_update ) {
 			return;
 		}
 
-		foreach ( $submenu[ $slug ] as $i => $submenu_item ) {
-			if ( array_key_exists( $submenu_item[2], $submenus_to_update ) ) {
-				$submenu_item[2] = $submenus_to_update[ $submenu_item[2] ];
-				// phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
-				$submenu[ $slug ][ $i ] = $submenu_item;
+		/**
+		 * Iterate over all submenu items and add the hide the submenus with CSS classes.
+		 * This is done separately of the second foreach because the position of the submenu might change.
+		 */
+		foreach ( $submenu[ $slug ] as $index => $item ) {
+			if ( ! array_key_exists( $item[2], $submenus_to_update ) ) {
+				continue;
+			}
+
+			$css_classes = empty( $item[4] ) ? 'hide-if-js hide-if-no-js' : $item[4] . ' hide-if-js hide-if-no-js';
+
+			// phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+			$submenu [ $slug ][ $index ][4] = $css_classes;
+		}
+
+		$submenu_items = array_values( $submenu[ $slug ] );
+
+		/**
+		 * Iterate again over the submenu array. We need a copy of the array because add_submenu_page will add new elements
+		 * to submenu array that might cause an infinite loop.
+		 */
+		foreach ( $submenu_items as $i => $submenu_item ) {
+			if ( ! array_key_exists( $submenu_item[2], $submenus_to_update ) ) {
+				continue;
 			}
+
+			add_submenu_page(
+				$slug,
+				isset( $submenu_item[3] ) ? $submenu_item[3] : '',
+				isset( $submenu_item[0] ) ? $submenu_item[0] : '',
+				isset( $submenu_item[1] ) ? $submenu_item[1] : 'read',
+				$submenus_to_update[ $submenu_item[2] ],
+				'',
+				$i
+			);
```

Comment:
Can you explain why we need to hide the item and then re-add it? Also I think we wanted to avoid `add_submenu_page` and instead we were mutating the `$submenu`.

I will cc @mmtr who has more expertise than me here.

Reply:
There are third-party plugins that register hidden pages (like `aioseo-setup-wizard`) that expect WP Core submenu items to be available. If the built-in items like index.php are removed (or their slug changes, as we did before), the plugin's pages are generated with a different internal name, causing their URLs to break.

My initial approach was to avoid add_submenu_page as well because I thought there might some perf penalties and opted to use array_splice to update the array; however that meant that we won't preserve the initial array keys from $submenu. 

So, to avoid breaking any behaviour and to reduce my code complexity I thought that add_submenu_page was a better alternative than duplicating the WPCORE behavior. 

That's not set in stone, of course, we can change it if there are some side effects I'm not aware of :)

Reply:
Thanks for the detailed info. That makes perfect sense. What I cannot still understand is, since we are now only hiding them and not actually removing them, why do we need to add them back?

Reply:
We don't actually add them back, we add a new submenu item that has the same name and capabilities but we use a different slug (that it's passed in the $submenus_to_update array); that array looks something like this:
```
$submenus_to_update = [
    'old-slug-we-have' => 'our-new-url',
];
```
In line 198 we have something like $submenus_to_update[ $our_old_slug ], which means that the slug used in the new submenu item would be `our-new-url`.

Reply:
I think I can now understand, thanks for breaking this down. Since we may change the initial parent menu slug, we need to update that slug to the hidden submenus, right?

Reply:
Not really sure tbh, because the built-in pages are working as expected. I'm assuming that $menu is just for rendering and $submenu is the variable that's used for determining if a page exists or not, but that's just a guess.